keps/sig-api-machinery/4460-per-request-deadline/README.md

```diff
@@ -0,0 +1,916 @@
+
+- [create](https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go#L193-L218)
+- [delete](https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go#L128-L132)
+- [deletecollection](https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go#L270-L272)
+- [patch](https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go#L667-L689)
+- [update](https://github.com/kubernetes/kubernetes/blob/3a4c35cc89c0ce132f8f5962ce4b9a48fae77873/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/update.go#L225-L237)
+
+We can replace the asynchronous finisher with one that executes the request
+serially in the same goroutine as the invoker, as shown below:
+```
+func serialFinisher(ctx context.Context, fn ResultFunc) (runtime.Object, error) {
+	result := &result{}
+	func() {
+		// capture the panic here to be rethrown later, this is in
+		// keeping with the behavior of the asynchronous finisher
+		defer func() {
+			if reason := recover(); reason != nil {
+				// store the panic reason into the result.
+				result.reason = capture(reason)
+			}
+		}()
+		result.object, result.err = fn()
+	}()
+
+	return result.Return()
+}
+```
+
+If the feature gate `PerHandlerReadWriteTimeout` is enabled, we use the serial version.
+```
+func FinishRequest(ctx context.Context, fn ResultFunc) (runtime.Object, error) {
+	if utilfeature.DefaultFeatureGate.Enabled(features.PerHandlerReadWriteTimeout) {
+		return serialFinisher(ctx, fn)
+	}
+	return finishRequest(ctx, fn, postTimeoutLoggerWait, logPostTimeoutResult)
+}
+```
+
+
+### Monitor Post-Timeout Activity:
+
+Upon timeout of a request, the `timeout` filter today provides with the following:
+
+- increment the merics counter `request_post_timeout_total` by 1 for every 
+  request that has timed out.
+- print a line to the log
+
+```
+	err := fmt.Errorf("post-timeout activity - time-elapsed: %s, %v %q result: %v",
+		time.Since(timedOutAt), req.Method, req.URL.Path, result)
+	utilruntime.HandleError(err)
+```
+
+> After a request times out, the asynchronous finisher will spin up a new 
+> goroutine and wait up to `5m` for the callee/child goroutine to return.
+>
+> The timeout filter does the same thing, but waits indefinitely until the
+> callee goroutine returns.
+
+
+With the `timeout` filter removed, we need to build a new mechanism to maintain
+parity. We can setup a `PostStartHook` that will detect how long a requst handler
+goroutine is active after the timeout happens, and provide the same observability
+as today.
+
+One simple approach for alpha, is to maintain a global synchronized hashmap of 
```

Comment:
Timeouts are generally fairly infrequent thing, so I would like to avoid to constantly maintain the hashmap of all inflight requests.

Instead, would it be possible to:
- not touch the regular requests at all
- if we face RequestDeadline, we simply spin up a goroutine for it to track it post-timeout activity?

It starts a goroutine, but only for those requests that timeout (which is infrequent) and with very special purpose - I think it would generally work better than global tracking.

Reply:
Yes. Ideally we want to spin up goroutines only for requests that time out, but the tricky part is to get notified when the context expires from a request handler that is executing on a single goroutine. I will continue looking for a better option here. 

On the other hand, I think the global tracking will give us insights that we don't have today. We can introduce a debugging endpoint that dumps information about goroutines that are blocking, and for how long. We can also place a beta criteria - replace the global tracking with goroutines only for requests that time out.

Reply:
I share wojtek's concerns. I haven't devoted time to consider an alternative yet.

Reply:
@deads2k @wojtek-t  we could do something like this:
```
func(w http.ResponseWriter, req *http.Request) {
   deadline := started.Add(timeout)
   ctx, cancel := context.WithDeadline(ctx, deadline)
   defer cancel()

    stopFn := context.AfterFunc(ctx, func() {
	 // launch monitor goroutine
     })

    defer stopFn()
     handler.ServeHTTP(w, req)
}
```

for a request that returns in time, we are stopping the after func, and the monitor will not be executed for it. for requests that don't return before the context expires, the monitor will execute.
I am going to write a test for this.